src/pyqasm/maps/gates.py

```diff
@@ -1,1335 +1,1221 @@
-    UintType: int,
-    BoolType: bool,
-    FloatType: float,
-    ComplexType: complex,
-    # AngleType: None,  # not sure
-}
-
-# Reference: https://openqasm.com/language/types.html#allowed-casts
-VARIABLE_TYPE_CAST_MAP = {
-    BoolType: (int, float, bool, np.int64, np.float64, np.bool_),
-    IntType: (bool, int, float, np.int64, np.float64, np.bool_),
-    BitType: (bool, int, np.int64, np.bool_),
-    UintType: (bool, int, float, np.int64, np.uint64, np.float64, np.bool_),
-    FloatType: (bool, int, float, np.int64, np.float64, np.bool_),
-    AngleType: (float, np.float64),
-}
-
-ARRAY_TYPE_MAP = {
-    BitType: np.bool_,
-    IntType: np.int64,
-    UintType: np.uint64,
-    FloatType: np.float64,
-    ComplexType: np.complex128,
-    BoolType: np.bool_,
-    AngleType: np.float64,
-}
-
-# Reference : https://openqasm.com/language/types.html#arrays
-MAX_ARRAY_DIMENSIONS = 7
-
-# Reference : https://openqasm.com/language/classical.html#the-switch-statement
-# Paragraph 14
-SWITCH_BLACKLIST_STMTS = {
-    QubitDeclaration,
-    ClassicalDeclaration,
-    SubroutineDefinition,
-    QuantumGateDefinition,
-}
-
-SUPPORTED_QASM_VERSIONS = {"3.0", "3", "2", "2.0"}
-
-QUANTUM_STATEMENTS = (QuantumGate, QuantumBarrier, QuantumReset, QuantumMeasurementStatement)
+# Copyright (C) 2024 qBraid
+#
+# This file is part of pyqasm
+#
+# Pyqasm is free software released under the GNU General Public License v3
+# or later. You can redistribute and/or modify it under the terms of the GPL v3.
+# See the LICENSE file in the project root or <https://www.gnu.org/licenses/gpl-3.0.html>.
+#
+# THERE IS NO WARRANTY for pyqasm, as per Section 15 of the GPL v3.
+
+# pylint: disable=too-many-lines
+
+"""
+Module mapping supported QASM gates to lower level gate operations.
+
+"""
+
+
+from typing import Callable, Union
+
+import numpy as np
+from openqasm3.ast import (
+    FloatLiteral,
+    Identifier,
+    IndexedIdentifier,
+    QuantumGate,
+    QuantumPhase,
+)
+
+from pyqasm.elements import InversionOp
+from pyqasm.exceptions import ValidationError
+from pyqasm.linalg import kak_decomposition_angles
+from pyqasm.maps import OperatorFunction
+from pyqasm.maps.expressions import CONSTANTS_MAP
+
+
+OPERATOR_MAP: dict[str, OperatorFunction] = {
+    "+": lambda x, y: x + y,
+    "-": lambda x, y: x - y,
+    "*": lambda x, y: x * y,
+    "/": lambda x, y: x / y,
+    "%": lambda x, y: x % y,
+    "==": lambda x, y: x == y,
+    "!=": lambda x, y: x != y,
+    "<": lambda x, y: x < y,
+    ">": lambda x, y: x > y,
+    "<=": lambda x, y: x <= y,
+    ">=": lambda x, y: x >= y,
+    "&&": lambda x, y: x and y,
+    "||": lambda x, y: x or y,
+    "^": lambda x, y: x ^ y,
+    "&": lambda x, y: x & y,
+    "|": lambda x, y: x | y,
+    "<<": lambda x, y: x << y,
+    ">>": lambda x, y: x >> y,
+    "~": lambda x: ~x,
+    "!": lambda x: not x,
+    "UMINUS": lambda x: -x,
+}
+
+
+def qasm3_expression_op_map(op_name: str, *args) -> Union[float, int, bool]:

```

Comment:
similar as above, to be moved to `expressions.py`

Reply:
Done